crates/bevy_ecs/src/world/command_queue.rs

```diff
@@ -55,6 +61,58 @@ unsafe impl Send for CommandQueue {}
 unsafe impl Sync for CommandQueue {}
 
 impl CommandQueue {
+    /// Push a [`Command`] onto the queue.
+    #[inline]
+    pub fn push<C>(&mut self, command: C)
+    where
+        C: Command,
+    {
+        // SAFETY: CommandQueue has a unique mutable borrow
+        unsafe { self.get_raw() }.push(command)
+    }
+    /// Execute the queued [`Command`]s in the world after applying any commands in the world's internal queue.
+    /// This clears the queue.
+    #[inline]
+    pub fn apply(&mut self, world: &mut World) {
+        // flush the previously queued entities
+        world.flush_entities();
+
+        // flush the world's internal queue
+        world.flush_commands();
+
+        // SAFETY: A reference is always a valid pointer
+        unsafe {
+            self.get_raw().apply_or_drop_queued(Some(world));
+        }
+    }
+
+    /// Take all commands from `other` and append them to `self`, leaving `other` empty
+    pub fn append(&mut self, other: &mut CommandQueue) {
+        self.bytes.append(&mut other.bytes);
+    }
+
+    /// Returns false if there are any commands in the queue
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.cursor >= self.bytes.len()
+    }
+
+    pub unsafe fn get_raw(&mut self) -> RawCommandQueue {
```

Comment:
This is unsafe to bubble up the safety requirements you mention in your other comment, the invariant for the struct is it must always point to valid data, that way pushing to it is always valid. Since this is all (now) private to the crate their shouldn't be any way for users to violate this.